src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -25978,76 +26183,1350 @@ class COblivious : IEnumerable
+                {
+                    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => throw null!;
+                    IEnumerator<T> IEnumerable<T>.GetEnumerator() => throw null!;
+                    public void Add<U>(U u) where U : notnull { }
+                }
+                """;
+
+            // We don't analyze the Add methods
+            CreateCompilation(source).VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void ElementNullability_InterfaceTypeCollection()
+        {
+            string src = """
+                using System.Collections.Generic;
+
+                #nullable enable
+                IEnumerable<string?> x1 = [null];
+                IEnumerable<string> x2 = [null];
+
+                #nullable disable
+                IEnumerable<string>
+                #nullable enable
+                    x3 = [null];
+                """;
+
+            CreateCompilation(src, targetFramework: TargetFramework.Net80).VerifyEmitDiagnostics(
+                // (5,27): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                // IEnumerable<string> x2 = [null];
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(5, 27)
+                );
+        }
+
+        [Fact]
+        public void ElementNullability_NoConversion()
+        {
+            string src = """
+                #nullable enable
+                int[] x = [null];
+                int y = null;
+                """;
+
+            CreateCompilation(src).VerifyEmitDiagnostics(
+                // (2,12): error CS0037: Cannot convert null to 'int' because it is a non-nullable value type
+                // int[] x = [null];
+                Diagnostic(ErrorCode.ERR_ValueCantBeNull, "null").WithArguments("int").WithLocation(2, 12),
+                // (3,9): error CS0037: Cannot convert null to 'int' because it is a non-nullable value type
+                // int y = null;
+                Diagnostic(ErrorCode.ERR_ValueCantBeNull, "null").WithArguments("int").WithLocation(3, 9)
+                );
+        }
+
+        [Fact]
+        public void ElementNullability_CheckExpressions()
+        {
+            string src = """
+                #nullable enable
+                object? o = null;
+                string[] x = [o.ToString()];
+                """;
+
+            CreateCompilation(src).VerifyEmitDiagnostics(
+                // (3,15): warning CS8602: Dereference of a possibly null reference.
+                // string[] x = [o.ToString()];
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "o").WithLocation(3, 15)
                 );
         }
 
+        [Fact]
+        public void ElementNullability_Inference_MaybeNullElement()
+        {
+            string src = """
+                using System.Collections.Generic;
+
+                #nullable enable
+                string? element = null;
```

Comment:
>string? [](http://example.com/codeflow?start=16&length=7)

Consider using `string` rather than `string?` so it's clear the method type inference is based on the inferred rather than declared nullability.